File: ios/Pushdy.m
//#import "Pushdy.h"

//@implementation Pushdy
//
//RCT_EXPORT_MODULE()

//RCT_EXPORT_METHOD(sampleMethod:(NSString *)stringArgument numberParameter:(nonnull NSNumber *)numberArgument callback:(RCTResponseSenderBlock)callback)
//{
//    // TODO: Implement some actually useful functionality
//    callback(@[[NSString stringWithFormat: @"numberArgument22: %@ stringArgument: %@", numberArgument, stringArgument]]);
//}

//@end


/**
 * Change to use Swift instead of Objective-C
 * This guide might be useful for you:
 * https://teabreak.e-spres-oh.com/swift-in-react-native-the-ultimate-guide-part-1-modules-9bb8d054db03#f662
 */
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(Pushdy, NSObject)

RCT_EXTERN_METHOD(
  sampleMethod: (NSString *)stringArgument
  numberArgument: (nonnull NSNumber *)numberArgument
  resolve: (RCTPromiseResolveBlock)resolve
  rejecter: (RCTPromiseRejectBlock)reject
)

@end

File: ios/Pushdy.swift
//
//  Pushdy.swift
//  react-native-pushdy
//
//  Created by luatnd on 16/12/19.
//

import Foundation
import os
import PushdySDK


@objc(Pushdy)
class RNPushdy: NSObject {
    // private var osLogEnabled = false

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc
    func sampleMethod(_
        stringArgument: String, numberArgument: NSNumber,
        resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock
    ) -> Void {
        let msg:String = "stringArgument: \(stringArgument), numberArgument: \(numberArgument)"

        NSLog(msg);
        
        if (Int(truncating: numberArgument) < 0) {
            reject("InvalidNumberArgument", "numberArgument cannot be negative", NSError(domain: "", code: 200, userInfo: nil))
        } else {
            // let a:Array<Any> = [msg, Int(truncating: numberArgument) * 2];
            resolve([msg, Int(truncating: numberArgument) * 2])
        }
    }

//    @objc
//    func constantsToExport() -> [AnyHashable : Any]! {
//        return ["initialCount": 0]
//    }
//    // This function is not completed :(
//    func safeOsLog(msg: StaticString, type: String, _ args: CVarArg...) {
//        if #available(iOS 10.0, *) {
//            var logType = OSLogType.default
//            switch type {
//                case "info":
//                    logType = .info
//                    break
//                case "debug":
//                    logType = .debug
//                    break
//                case "error":
//                    logType = .error
//                    break
//                case "fault":
//                    logType = .fault
//                    break
//                default:
//                    logType = OSLogType.default
//            }
//
//            os_log(msg, log: OSLog.default, type: logType, args)
//        }
//    }
}
